### ISSUE TITLE
Ease In/Out Speed

### ISSUE THREAD

**Issue by camjac251** [WheteThunger/BetterElevators#1]
Is it possible to ease the stopping motion of the elevator? I wanted to have a speed of 25 but as soon as someone comes close to reaching the top or bottom, it stops suddenly when it's reached.

Is it possible to support a bézier like curve to emulate a braking system that slows down the movement toward the end?

I also had a question about the `SpeedIncreasePerFloor` variable if you don't mind. Is this a single multiplier that applies across all of the floors? Is it a gradual one that multiplies every floor as its passing? Like `speed/floor`, `1/1`-`1.5/2`-`2/3`-`2.5/4`-`3/5`. Or is it like `1/1`-`3/2`-`3/3`-`3/4`-`3/5`?

**Comment by WheteThunger** [comment]
That would be the ideal experience, but so far this plugin is just using the same LeanTween utility which Rust already uses for vanilla Elevator movement. It has some different types that I could experiment with for better easing. Also possible to implement something custom, but it would take some work, so possibly not worth it since it will primarily benefit people who are using super speed. Any changes to speed like this will complicate maintaining speed when the elevator height changes, but that's probably an acceptable problem as it's rare. Counter updates may also become trickier.

To answer your question about `SpeedIncreasePerFloor`, the formula is:
```csharp
BaseSpeed + (FloorsToMove - 1) * SpeedIncreasePerFloor
```

So your first example (`1/1`-`1.5/2`-`2/3`-`2.5/4`-`3/5`) looks right.

**Comment by camjac251** [comment]
Maybe my example was a bit flawed. Let's say you want to go up 100 rows of elevators (I'm making a large tower in RustEdit and was hoping to have the player to reach the top in less than 10 seconds, vs a few minutes without any speed adjustment).

Would the speed stay constant if started from floor one and used the "to top" function? Could it be possible for gradual change like with every floor you pass, the speed gets multiplied. I think right now all of them are given the same formula right?

In the example above I was meaning if you start at floor 1 and hit "to top" on a 5 row elevator, the first one gradually increases speed with each row that you pass while the second one applies the same speed across all floors if the formula is for the entire movement instead of varied

**Comment by WheteThunger** [comment]
Yes, ths speed is constant from start to end. The speed is calculated when the lift starts moving, based on the distance that is going to be travelled. Currently the plugin just calculates the speed and sets it on the elevator, and the game uses a function to perform the animation with constant speed.

I was experimenting with other parameters to the game's built-in animation function. There is a cubic easing option that was reasonable. I'm only able to pass it the desired time, though, so the next problem is determining a formula to determine the right time to calculate, since the issue I was seeing was that if I am going farther, then the initial acceleration would actually be slower.

I prefer going the route of the built-in utility rather than implementing something custom, but we'll see.

**Comment by WheteThunger** [comment]
I pushed some changes to the develop branch.
https://github.com/WheteThunger/BetterElevators/tree/develop

The old speed presets should still work, but the new default configuration includes quadratic and cubic easing presets. You can achieve 10 seconds for 100 floors with either 2x quadratic or 1x cubic.

Let me know how that is working out.

**Comment by camjac251** [comment]
Oh wow this is incredible, thank you! I really like these settings for 100 floors.
```json
  "DefaultSpeed": {
    "BaseSpeed": 0.8,
    "MaxSpeed": 100,
    "EaseType": "Quadratic"
  },
```
I like the idea of an elevator starting off slow and gaining momentum as it goes from floor 1 to 100, and without `MaxSpeed`, it isn't able to reach as fast without forcing the starting speed to be fast too. Would it be possible to not deprecate this function?

I decided to try out the counter floor option too and it's amazing. Although it might not work with my use case. I'm designing a map in RustEdit and it isn't possible to do the same in that editor so I'd be forced to making it ingame every wipe just to get those counters. I was curious if maybe Zone Manager could help but it would become a different plugin entirely then.

**Comment by WheteThunger** [comment]
Keeping BaseSpeed would basically require re-implementing the easing from scratch, unfortunately. The FacePunch utility just allows passing the distance and desired time, and the easing function ultimately determines the acceleration and max speed. It also doesn't really support a custom easing function. If 1.6 quadratic is too fast initially then you can try cubic with a lower base speed which will accelerate more quickly and reach a higher top speed. Right now I am just dividing the total time by BaseSpeed to give some speed control.

**Comment by WheteThunger** [comment]
Implemented acceleration/deceleration in v1.1.0. Currently available on the master branch and on the uMod website.
